torch/csrc/jit/mobile/nnc/context.cpp

```diff
@@ -0,0 +1,252 @@
+#include <torch/csrc/jit/mobile/nnc/context.h>
+
+#include <ATen/Functions.h>
+#include <ATen/core/functional.h>
+#include <c10/core/CPUAllocator.h>
+
+#include <torch/csrc/jit/mobile/nnc/registry.h>
+
+namespace torch {
+namespace jit {
+namespace mobile {
+namespace nnc {
+
+constexpr int64_t kProducedNNCFileFormatVersion = 0x1L;
+
+namespace {
+
+c10::IValue Tup(std::vector<c10::IValue>&& ivalues) {
+  return c10::ivalue::Tuple::create(ivalues);
+}
+
+} // namespace
+
+InputSpec::InputSpec(const c10::IValue& value) {
+  auto dict = value.toGenericDict();
+  sizes_ = dict.at("sizes").toIntVector();
+  dtype_ = dict.at("dtype").toScalarType();
+}
+
+c10::IValue InputSpec::serialize() const {
+  c10::Dict<c10::IValue, c10::IValue> dict(
+      at::StringType::get(), at::AnyType::get());
+  dict.insert("sizes", sizes_);
+  dict.insert("dtype", dtype_);
+  return dict;
+}
+
+bool InputSpec::validate(const at::Tensor& input) const {
+  return input.sizes() == sizes_ && input.scalar_type() == dtype_;
+}
+
+OutputSpec::OutputSpec(const c10::IValue& value) {
+  auto dict = value.toGenericDict();
+  sizes_ = dict.at("sizes").toIntVector();
+  dtype_ = dict.at("dtype").toScalarType();
+}
+
+c10::IValue OutputSpec::serialize() const {
+  c10::Dict<c10::IValue, c10::IValue> dict(
+      at::StringType::get(), at::AnyType::get());
+  dict.insert("sizes", sizes_);
+  dict.insert("dtype", dtype_);
+  return dict;
+}
+
+at::Tensor OutputSpec::allocate() const {
+  return at::empty(
+      sizes_,
+      at::TensorOptions()
+          .dtype(dtype_)
+          .layout(at::kStrided)
+          .device(at::kCPU)
+          .requires_grad(false));
+}
+
+MemoryPlan::MemoryPlan(const c10::IValue& value) {
+  auto dict = value.toGenericDict();
+  buffer_sizes_ = dict.at("buffer_sizes").toIntVector();
+}
+
+c10::IValue MemoryPlan::serialize() const {
+  c10::Dict<c10::IValue, c10::IValue> dict(
+      at::StringType::get(), at::AnyType::get());
+  dict.insert("buffer_sizes", buffer_sizes_);
+  return dict;
+}
+
+void MemoryPlan::allocate(ExecutionState* state) const {
+  auto& allocations = state->preallocations_;
+  allocations.reserve(buffer_sizes_.size());
```

Comment:
Do you also need to clear? Not necessary here probably but if we call allocate multiple times.

Reply:
I'll add it.